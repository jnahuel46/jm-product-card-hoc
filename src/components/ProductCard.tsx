import styles from "../styles/styles.module.css";
import { useProduct } from "../hooks/useProduct";
import React, { createContext } from "react";
import {
  ProductContextProps,
  Product,
  OnChangeArgs,
  InitialValues,
  ProductCardHandlers,
} from "../interfaces/interfaces";

export const ProductContext = createContext({} as ProductContextProps);
export const { Provider } = ProductContext;

export interface Props {
  product: Product;
  //children?: ReactElement | ReactElement[];
  children: (args: ProductCardHandlers) => JSX.Element;
  className?: string;
  style?: React.CSSProperties;
  onChange?: (args: OnChangeArgs) => void;
  value?: number;
  initialValue?: InitialValues;
}

export const ProductCard = ({
  children,
  product,
  className,
  style,
  onChange,
  value,
  initialValue,
}: Props) => {
  const { counter, increaseBy, maxCount, reset, isMaxCountReaced } = useProduct(
    {
      onChange,
      product,
      value,
      initialValue,
    }
  );
  return (
    //Provee de contexto a todos los hijos del componente
    <Provider
      value={{
        counter,
        increaseBy,
        product,
        maxCount,
      }}
    >
      <div style={style} className={`${styles.productCard} ${className}`}>
        {/* Renderiza una funcion */}
        {children({
          count: counter,
          isMaxCountReached: isMaxCountReaced,
          maxCount: maxCount,
          product,
          increaseBy,
          reset,
        })}
      </div>
    </Provider>
  );
};
