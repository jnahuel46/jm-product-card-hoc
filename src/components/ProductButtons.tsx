import React, { useCallback, useContext } from "react";
import { ProductContext } from "./ProductCard";
import styles from "../styles/styles.module.css";
import { ButtonsProps } from "../interfaces/interfaces";

export const ProductButtons = ({ className, style }: ButtonsProps) => {
  const { increaseBy, counter, maxCount } = useContext(ProductContext);
  console.log(maxCount);
  const isMaxReached = useCallback(() => {
    if (maxCount === counter && !!maxCount) {
      return true
    } else {
      return false
    }
  }, [maxCount, counter]);

  //usar el use callback dep
  return (
    <div style={style} className={`${styles.buttonsContainer} ${className}`}>
      <button onClick={() => increaseBy(-1)} className={styles.buttonMinus}>
        -
      </button>
      <div className={styles.countLabel}> {counter} </div>
      <button
        onClick={() => increaseBy(+1)}
        className={`${styles.buttonAdd} ${isMaxReached() && styles.disabled }`}
      >
        +
      </button>
    </div>
  );
};
