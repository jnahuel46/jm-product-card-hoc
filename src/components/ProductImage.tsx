import React,{ useContext } from "react";
import { ProductContext } from "./ProductCard";
import noImages from "../assets/no-image.jpg";
import styles from "../styles/styles.module.css";
import { ImageProps } from "../interfaces/interfaces";

export const ProductImage = ({ img, className, style }: ImageProps) => {
  const { product } = useContext(ProductContext);

  let imgToShow: string;
  if (img) {
    imgToShow = img;
  } else if (product.img) {
    imgToShow = product.img;
  } else {
    imgToShow = noImages;
  }

  return (
    <img
      style={style}
      className={`${styles.productImg} ${className}`}
      src={imgToShow}
      alt="Coffe Mug"
    />
  );
};
