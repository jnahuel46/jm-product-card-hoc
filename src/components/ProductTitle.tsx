import React, { useContext } from "react";
import { ProductContext } from "./ProductCard";
import styles from "../styles/styles.module.css";
import { TitleProps } from "../interfaces/interfaces";

export const ProductTitle = ({ title, className }: TitleProps) => {
  const { product } = useContext(ProductContext);

  return (
    <span className={`${styles.productDescription} ${className}`}>
      {title ? title : product.title}
    </span>
  );
};
