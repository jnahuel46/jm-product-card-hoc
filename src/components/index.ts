import { ProductCard as ProductCardHoc } from "./ProductCard";

import { ProductButtons } from "./ProductButtons";
import { ProductImage } from "./ProductImage";
import { ProductTitle } from "./ProductTitle";
import { ProductHocMain } from "../interfaces/interfaces";

export { ProductButtons } from "./ProductButtons";
export { ProductImage } from "./ProductImage";
export { ProductTitle } from "./ProductTitle";

export const ProductCard: ProductHocMain = Object.assign(ProductCardHoc, {
  Title: ProductTitle,
  Image: ProductImage,
  Buttons: ProductButtons,
});

export default ProductCard;
