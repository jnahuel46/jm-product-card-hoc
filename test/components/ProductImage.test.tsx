import React from 'react';
import renderer from 'react-test-renderer';
import { ProductImage, ProductCard } from '../../src/components';
import { product1 } from '../data/products';
import { product2 } from '../data/products';

describe('ProductImage', () => {
  test('should show chorrectly the component without img', () => {
    const wrapper = renderer.create(
      <ProductImage img={'https://random-test-img.jpg'} className="class-test" />
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
  test('should show chorrectly the component with the img of the product', () => {
    const wrapper = renderer.create(
      <ProductCard product={product1}>
        {() => <ProductImage img={product2.img} />}
      </ProductCard>
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
