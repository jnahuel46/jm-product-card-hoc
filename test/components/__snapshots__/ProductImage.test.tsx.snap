// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImage should show chorrectly the component with the img of the product 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Coffe Mug"
    className="productImg undefined"
    src="./coffee-mug.png"
  />
</div>
`;

exports[`ProductImage should show chorrectly the component without img 1`] = `
<img
  alt="Coffe Mug"
  className="productImg class-test"
  src="https://random-test-img.jpg"
/>
`;
